File: kotlin/src/test/kotlin/com/bhengubv/circleai/memory/sync/SyncBridgesTest.kt
// SyncBridgesTest.kt
//
// Verifies the three sync bridges against the C# reference:
//   PersonaStateSyncBridge      — save persists + pushes; encode/decode round-trip
//   LoraAdapterSyncBridge       — publish reads file + pushes; tryWrite decodes + writes bytes
//   CompanionConversationSyncBridge — publish/terminate + tryDecode

package com.bhengubv.circleai.memory.sync

import com.bhengubv.circleai.memory.IPersonaStore
import com.bhengubv.circleai.memory.PersonaState
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.nio.file.Path
import java.time.Instant
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class SyncBridgesTest {

    private class FakePersonaStore : IPersonaStore {
        val saved = ArrayList<PersonaState>()
        override suspend fun loadAsync(userId: String): PersonaState = PersonaState(userId)
        override suspend fun saveAsync(persona: PersonaState) { saved.add(persona) }
    }

    private fun freshEngine(nodeId: String = "A"): Pair<CompanionStateSyncEngine, ISyncableEntryStore> {
        val hub = InProcessSyncHub()
        val channel = InProcessCompanionStateChannel(hub, nodeId)
        val store = InMemorySyncableEntryStore()
        val clock = HybridLogicalClock(1) { 1000L }
        val engine = CompanionStateSyncEngine(channel, store, clock) { Instant.EPOCH }
        return engine to store
    }

    // -- PersonaStateSyncBridge ---------------------------------------------

    @Test
    fun `persona save persists locally and pushes an entry`() = runTest {
        val (engine, store) = freshEngine()
        engine.start()
        val fakeStore = FakePersonaStore()
        val bridge = PersonaStateSyncBridge(fakeStore, engine)

        val persona = PersonaState("user-7").apply {
            verbosity = "brief"
            formality = "formal"
            preferredLocale = "en-ZA"
            topicWeights["finance"] = 2.5f
            disfavouredTopics.add("sport")
            totalInteractions = 12
            positiveSignals = 9
            negativeSignals = 3
        }

        bridge.save(persona)
        assertEquals(1, fakeStore.saved.size)
        val entry = store.get(PersonaStateSyncBridge.EntityType, "user-7")
        assertNotNull(entry)

        val decoded = PersonaStateSyncBridge.tryDecode(entry!!)
        assertNotNull(decoded)
        assertEquals("user-7", decoded!!.userId)
        assertEquals("brief", decoded.verbosity)
        assertEquals("formal", decoded.formality)
        assertEquals("en-ZA", decoded.preferredLocale)
        assertEquals(2.5f, decoded.topicWeights["finance"])
        assertTrue(decoded.disfavouredTopics.contains("sport"))
        assertEquals(12, decoded.totalInteractions)
        assertEquals(9, decoded.positiveSignals)
        assertEquals(3, decoded.negativeSignals)

        engine.close()
    }

    @Test
    fun `persona tryDecode rejects tombstone and wrong type`() = runTest {
        val tombstone = SyncableEntry(
            PersonaStateSyncBridge.EntityType, "u", 1, true, "h", "", "n", Instant.EPOCH,
        )
        assertNull(PersonaStateSyncBridge.tryDecode(tombstone))

        val wrongType = SyncableEntry("Other", "u", 1, false, "h", "{}", "n", Instant.EPOCH)
        assertNull(PersonaStateSyncBridge.tryDecode(wrongType))
    }

    // -- LoraAdapterSyncBridge ----------------------------------------------

    @Test
    fun `lora publish reads file and tryWrite round-trips the bytes`(@TempDir dir: Path) = runTest {
        val (engine, store) = freshEngine()
        engine.start()
        val bridge = LoraAdapterSyncBridge(engine)

        val src = File(dir.toFile(), "adapter.bin")
        val bytes = byteArrayOf(1, 2, 3, 4, 5, 9, 8, 7)
        src.writeBytes(bytes)

        bridge.publish("personal-u1", src.absolutePath, stepCount = 42)
        val entry = store.get(LoraAdapterSyncBridge.EntityType, "personal-u1")
        assertNotNull(entry)

        val dest = File(dir.toFile(), "nested/out.bin")
        val snapshot = LoraAdapterSyncBridge.tryWrite(entry!!, dest.absolutePath)
        assertNotNull(snapshot)
        assertEquals("personal-u1", snapshot!!.adapterId)
        assertEquals(42L, snapshot.stepCount)
        assertTrue(dest.exists())
        assertContentEquals(bytes, dest.readBytes())

        engine.close()
    }

    @Test
    fun `lora tryWrite rejects tombstone and wrong type`() = runTest {
        val tombstone = SyncableEntry(
            LoraAdapterSyncBridge.EntityType, "a", 1, true, "h", "", "n", Instant.EPOCH,
        )
        assertNull(LoraAdapterSyncBridge.tryWrite(tombstone, "ignored"))
        val wrongType = SyncableEntry("Other", "a", 1, false, "h", "{}", "n", Instant.EPOCH)
        assertNull(LoraAdapterSyncBridge.tryWrite(wrongType, "ignored"))
    }

    @Test
    fun `lora publish rejects missing file`() = runTest {
        val (engine, _) = freshEngine()
        engine.start()
        val bridge = LoraAdapterSyncBridge(engine)
        var threw = false
        try {
            bridge.publish("id", "C:/nonexistent/definitely/missing.bin", 1)
        } catch (_: java.io.FileNotFoundException) {
            threw = true
        }
        assertTrue(threw)
        engine.close()
    }

    // -- CompanionConversationSyncBridge ------------------------------------

    @Test
    fun `conversation publish and tryDecode round-trip`() = runTest {
        val (engine, store) = freshEngine()
        engine.start()
        val bridge = CompanionConversationSyncBridge(engine)

        val delta = ConversationStateDelta(
            sessionId = "sess-1",
            userText = "what's the weather",
            assistantText = "Checking",
            isTurnComplete = false,
            startedAtUtc = Instant.EPOCH.toString(),
            updatedAtUtc = Instant.EPOCH.plusSeconds(2).toString(),
        )
        bridge.publish(delta)

        val entry = store.get(CompanionConversationSyncBridge.EntityType, "sess-1")
        assertNotNull(entry)
        val decoded = CompanionConversationSyncBridge.tryDecode(entry!!)
        assertNotNull(decoded)
        assertEquals("sess-1", decoded!!.sessionId)
        assertEquals("what's the weather", decoded.userText)
        assertEquals("Checking", decoded.assistantText)
        assertEquals(false, decoded.isTurnComplete)

        engine.close()
    }

    @Test
    fun `conversation terminate writes a tombstone`() = runTest {
        val (engine, store) = freshEngine()
        engine.start()
        val bridge = CompanionConversationSyncBridge(engine)
        bridge.publish(
            ConversationStateDelta("sess-2", "hi", "", true, Instant.EPOCH.toString(), Instant.EPOCH.toString())
        )
        bridge.terminate("sess-2")
        val entry = store.get(CompanionConversationSyncBridge.EntityType, "sess-2")
        assertNotNull(entry)
        assertTrue(entry!!.isTombstone)
        assertNull(CompanionConversationSyncBridge.tryDecode(entry))
        engine.close()
    }
}

File: go/tests/sync_bridges_test.go
// sync_bridges_test.go
//
// Verifies the three sync bridges (ported from PersonaStateSyncBridge.cs,
// LoraAdapterSyncBridge.cs, CompanionConversationSyncBridge.cs):
//   - PersonaStateSyncBridge.Save persists locally AND pushes a decodable entry.
//   - LoraAdapterSyncBridge.Publish base64-encodes a file; TryWrite round-trips
//     the bytes to a destination path.
//   - CompanionConversationSyncBridge publishes deltas and terminates via
//     tombstone; TryDecode round-trips the delta and rejects tombstones.

package circleai_test

import (
	"context"
	"os"
	"path/filepath"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

// wireBridgeEngine builds an engine over a fresh in-process channel + store.
func wireBridgeEngine(t *testing.T, nodeID string) (*circleai.CompanionStateSyncEngine, *circleai.InMemorySyncableEntryStore, func()) {
	t.Helper()
	hub := circleai.NewInProcessSyncHub()
	ch, err := circleai.NewInProcessCompanionStateChannel(hub, nodeID)
	if err != nil {
		t.Fatalf("channel: %v", err)
	}
	store := circleai.NewInMemorySyncableEntryStore()
	eng, err := circleai.NewCompanionStateSyncEngine(ch, store, circleai.MustNewHybridLogicalClock(1), nil)
	if err != nil {
		t.Fatalf("engine: %v", err)
	}
	return eng, store, func() { ch.Close() }
}

func TestPersonaStateSyncBridge_SaveAndDecode(t *testing.T) {
	ctx := context.Background()
	eng, store, done := wireBridgeEngine(t, "A")
	defer done()
	personaStore := circleai.NewInMemoryPersonaStore()

	bridge, err := circleai.NewPersonaStateSyncBridge(personaStore, eng)
	if err != nil {
		t.Fatalf("bridge: %v", err)
	}

	loc := "en-ZA"
	persona := circleai.NewPersonaState("user-42")
	persona.Verbosity = "brief"
	persona.Formality = "casual"
	persona.PreferredLocale = &loc
	persona.TopicWeights["finance"] = 2.5
	persona.DisfavouredTopics["gossip"] = struct{}{}
	persona.PositiveSignals = 7

	if err := bridge.Save(ctx, persona); err != nil {
		t.Fatalf("Save: %v", err)
	}

	// Persisted locally.
	loaded, _ := personaStore.Load(ctx, "user-42")
	if loaded.Verbosity != "brief" {
		t.Errorf("persona store: got verbosity %q", loaded.Verbosity)
	}

	// Pushed as a decodable syncable entry.
	entry, _ := store.Get(ctx, circleai.PersonaStateEntityType, "user-42")
	if entry == nil {
		t.Fatal("persona not pushed to sync store")
	}
	decoded, ok := circleai.TryDecodePersonaState(*entry)
	if !ok {
		t.Fatal("decode failed")
	}
	if decoded.UserID != "user-42" || decoded.Verbosity != "brief" || decoded.Formality != "casual" {
		t.Errorf("decoded mismatch: %+v", decoded)
	}
	if decoded.PreferredLocale == nil || *decoded.PreferredLocale != "en-ZA" {
		t.Errorf("locale round-trip: %v", decoded.PreferredLocale)
	}
	if decoded.TopicWeights["finance"] != 2.5 {
		t.Errorf("topic weight round-trip: %v", decoded.TopicWeights)
	}
	if _, ok := decoded.DisfavouredTopics["gossip"]; !ok {
		t.Errorf("disfavoured round-trip: %v", decoded.DisfavouredTopics)
	}
	if decoded.PositiveSignals != 7 {
		t.Errorf("positive signals round-trip: %d", decoded.PositiveSignals)
	}
}

func TestLoraAdapterSyncBridge_PublishAndWrite(t *testing.T) {
	ctx := context.Background()
	eng, store, done := wireBridgeEngine(t, "A")
	defer done()

	bridge, err := circleai.NewLoraAdapterSyncBridge(eng)
	if err != nil {
		t.Fatalf("bridge: %v", err)
	}

	dir := t.TempDir()
	src := filepath.Join(dir, "adapter.bin")
	original := []byte{0x00, 0x01, 0x02, 0xDE, 0xAD, 0xBE, 0xEF}
	if err := os.WriteFile(src, original, 0o644); err != nil {
		t.Fatalf("write src: %v", err)
	}

	if err := bridge.Publish(ctx, "personal-42", src, 128); err != nil {
		t.Fatalf("Publish: %v", err)
	}

	entry, _ := store.Get(ctx, circleai.LoraAdapterEntityType, "personal-42")
	if entry == nil {
		t.Fatal("adapter not pushed")
	}

	dst := filepath.Join(dir, "nested", "out.bin")
	snap, ok, err := circleai.TryWriteLoraAdapter(*entry, dst)
	if err != nil {
		t.Fatalf("TryWrite: %v", err)
	}
	if !ok {
		t.Fatal("TryWrite returned not-ok for a valid adapter entry")
	}
	if snap.AdapterID != "personal-42" || snap.StepCount != 128 {
		t.Errorf("snapshot fields: %+v", snap)
	}
	written, err := os.ReadFile(dst)
	if err != nil {
		t.Fatalf("read dst: %v", err)
	}
	if string(written) != string(original) {
		t.Errorf("bytes round-trip mismatch: got %v want %v", written, original)
	}

	// Wrong entity type → not ok.
	other := circleai.SyncableEntry{EntityType: "PersonaState", Payload: "{}"}
	if _, ok, _ := circleai.TryWriteLoraAdapter(other, dst); ok {
		t.Error("non-adapter entry should not decode as adapter")
	}
	// Tombstone → not ok.
	tomb := circleai.SyncableEntry{EntityType: circleai.LoraAdapterEntityType, IsTombstone: true}
	if _, ok, _ := circleai.TryWriteLoraAdapter(tomb, dst); ok {
		t.Error("tombstone adapter entry should not decode")
	}
}

func TestCompanionConversationSyncBridge_PublishTerminateDecode(t *testing.T) {
	ctx := context.Background()
	eng, store, done := wireBridgeEngine(t, "A")
	defer done()

	bridge, err := circleai.NewCompanionConversationSyncBridge(eng)
	if err != nil {
		t.Fatalf("bridge: %v", err)
	}

	started := time.Date(2026, 7, 8, 10, 0, 0, 0, time.UTC)
	delta := circleai.ConversationStateDelta{
		SessionID:      "sess-1",
		UserText:       "hey B",
		AssistantText:  "Hi!",
		IsTurnComplete: false,
		StartedAtUTC:   started,
		UpdatedAtUTC:   started.Add(2 * time.Second),
	}
	if err := bridge.Publish(ctx, delta); err != nil {
		t.Fatalf("Publish: %v", err)
	}

	entry, _ := store.Get(ctx, circleai.ConversationStateEntityType, "sess-1")
	if entry == nil {
		t.Fatal("delta not pushed")
	}
	decoded, ok := circleai.TryDecodeConversationDelta(*entry)
	if !ok {
		t.Fatal("decode failed")
	}
	if decoded.UserText != "hey B" || decoded.AssistantText != "Hi!" || decoded.IsTurnComplete {
		t.Errorf("decoded mismatch: %+v", decoded)
	}
	if !decoded.StartedAtUTC.Equal(started) {
		t.Errorf("started round-trip: got %v want %v", decoded.StartedAtUTC, started)
	}

	// Terminate → tombstone; decode must reject it.
	if err := bridge.Terminate(ctx, "sess-1"); err != nil {
		t.Fatalf("Terminate: %v", err)
	}
	tomb, _ := store.Get(ctx, circleai.ConversationStateEntityType, "sess-1")
	if tomb == nil || !tomb.IsTombstone {
		t.Fatalf("terminate should tombstone: %+v", tomb)
	}
	if _, ok := circleai.TryDecodeConversationDelta(*tomb); ok {
		t.Error("tombstone should not decode to a delta")
	}
}

func TestSyncBridges_ConstructorValidation(t *testing.T) {
	if _, err := circleai.NewPersonaStateSyncBridge(nil, nil); err == nil {
		t.Error("nil deps should error")
	}
	if _, err := circleai.NewLoraAdapterSyncBridge(nil); err == nil {
		t.Error("nil engine should error")
	}
	if _, err := circleai.NewCompanionConversationSyncBridge(nil); err == nil {
		t.Error("nil engine should error")
	}
}
